<p>
<font size='5' face='Georgia, Arial'>IIC-2233 Apunte Programación Avanzada</font><br>
<font size='1'> Material creado por Equipo Docente IIC2233. Modificado en 2021-1 al 2023-1 por Equipo Docente IIC2233</font>
</p>

# Tabla de contenidos

1. [API (*Application Programming Interface*)](#API-(Application-Programming-Interface))
2. [HTTP (***Hypertext Transfer Protocol***)](#HTTP-(Hypertext-Transfer-Protocol))
3. [*Client-side script*](#Client-side-script)
    1. [Ejemplo consumiendo una API (método get)](#Ejemplo-consumiendo-una-API-(método-get))
    2. [Retorno de una API](#Retorno-de-una-API)
    3. [Uso de `post`](#Uso-de-post)
    4. [Autenticación en _headers_](#Autenticación-en-headers)
    5. [Caso aplicado: API de GITHUB](#Caso-aplicado:-API-de-GITHUB)
4. [*Server-side App*](#Server-side-App)

Durante el capítulo de *networking* aprendimos el uso de *sockets* y algunos protocolos para establecer la comunicación a través de una arquitectura cliente-servidor. En este capítulo, revisaremos la comunicación entre dos dispositivos mediante la **web**.

***Web services*** es el conjunto de aplicaciones cliente-servidor que se comunican a través de la web mediante un protocolo diseñado para ello. Podemos ver este tipo de servicios como aplicaciones que pueden ser accedidas por otras aplicaciones a través de una red de computadores (internet).

Por ejemplo, cuando nuestro navegador (cliente) consume un sitio web (servidor), por cada **llamada al servidor**, una aplicación escrita en algún lenguaje de programación **envía una respuesta** en [HTML](https://es.wikipedia.org/wiki/HTML) (el lenguaje que se utiliza para definir la estructura de un sitio web) para que nuestro navegador la despliegue. Los _web services_ funcionan de forma similar, donde la salida está dirigida a una **aplicación que consume** esta información. Para que la comunicación sea posible, el formato de los mensajes debe ser conocido por ambas partes, para que la información pueda ser interpretada correctamente.


## API (*Application Programming Interface*)

En general se conoce como **API** al conjunto de funciones que son expuestas por un servicio para ser utilizadas por otros programas. Podemos ver al servicio como una clase, y a la API como el conjunto de métodos de esa clase. El servicio puede ser un *web service* o cualquier paquete que exponga una interfaz, por ejemplo una librería de Python. Sin embargo, en este capítulo cuando hablemos de API nos estaremos refiriendo a los métodos expuestos por un servicio web (un *web service*).

Una gran parte de los servicios actuales exponen una API, y permiten que otras aplicaciones se conectan a ellas. De esta manera podemos construir aplicaciones que utilizan servicios que se encuentren en computadores remotos, e interactuar con ellos.

En una red de computadores, cada página web de internet es almacenada en un computador remoto que ejecuta un proceso servidor. Un servidor remoto es simplemente un programa que escuchas *requests* (solicitudes) y envía *responses* (respuestas) de acuerdo a un protocolo.

Tú puedes utilizar tu propio computador para servir un sitio web. De hecho, los desarrolladores de *software* usan sus propios computadores como servidores locales al crear sitios web antes de publicarlos al mundo.

Cuando escribes https://www.facebook.com en tu navegador, este envía una *request* a un servidor remoto de Facebook. Una vez que tu navegador recibe la respuesta del servidor, este la interpreta y despliega una página para ti.

Para el navegador (cliente), el servidor de Facebook es una API. Esto significa que cada vez que tú visitas una página en la web, tú interactúas con alguna API en un servidor remoto. Una API no es lo mismo que un servidor remoto, pero es la parte de este que recibe las *requests* y envía *responses* (respuestas).

Si cada objeto tecnológico de tu casa expusiera una API, podrías controlarla completamente desde tu celular o desde cualquier programa en Python u otro lenguaje (ver más en [internet de las cosas](https://es.wikipedia.org/wiki/Internet_de_las_cosas)).


## HTTP (***Hypertext Transfer Protocol***)

Gran parte de las arquitecturas de *web services* se basan en el uso del protocolo **HTTP**. Este protocolo de aplicación está encargado de proporcionar una capa para realización de transacciones y así permitir la comunicación entre clientes y servidores. HTTP trabaja como un protocolo ***request-response*** en donde el cliente hace una solicitud (*request*) y el servidor responde con la información solicitada (*response*).

HTTP es un protocolo en el que el servidor no guarda ninguna información de las conexiones. Por ejemplo, al acceder a métodos de un servicio web que requiere identificación del cliente, este deberá en cada consulta enviar **algo** que acredite su identidad.

El funcionamiento de este protocolo se basa en la definición de métodos o verbos que indican la acción a desarrollar por un determinado recurso. Los recursos pueden ser datos existentes en el servidor (por ejemplo, archivos) o bien una salida generada dinámicamente. La versión HTTP/1.1 incluye **cinco** métodos:

- `GET`: recupera una representación de un recurso sin cambiar nada en el servidor.
- `POST`: crea un recurso.
- `PATCH`: aplica modificaciones parciales a un recurso.
- `PUT`: reemplaza completamente un recurso existente.
- `DELETE`: elimina un recurso.

HTTP incluye también un conjunto de códigos de estado mediante los cuales se entrega información al cliente sobre el resultado de su petición. Algunos códigos comunes de respuesta son:

- `200` : OK. Solicitud exitosa.
- `403` : Prohibido. La petición es aceptada, pero el servidor rechaza responderla.
- `404` : No encontrado. El recurso solicitado no ha sido encontrado.
- `500` : Error interno del servidor.

Para más detalle de los códigos pueden revisar el siguiente [enlace]( http://www.w3schools.com/tags/ref_httpmessages.asp).

La siguiente figura muestra un ejemplo con la estructura de los mensaje HTTP para la *request* del cliente y para la *response* desde el servidor.

![](imgs/http_message.png)


## *Client-side script*

En esta sección veremos, desde el punto de vista del cliente, cómo efectuar *requests* a un servidor que mantiene un servicio web. En Python, la librería `requests` nos permite interactuar con servicios disponibles en algún *web service*. La librería, además, integra los métodos para serialización en JSON.

Para instalar la librería `requests`, en cualquier terminal debes correr el comando `pip3 install requests`. Otra opción, es ejecutar la siguiente celda, que instalará la librería en el mismo entorno en que estés corriendo este jupyter (de todas maneras recomendamos instalarlo desde la consola):


In [ ]:
!pip3 install requests

Para generar una petición mediante `GET` usamos el método `get(url)` que recibe por argumento el llamado al recurso.


In [2]:
import requests

# Esta url contiene la dirección del web service 
# y los parámetros que se requieren para la consulta
url = 'https://api.github.com/repos/IIC2233/Syllabus/issues/8'
response = requests.get(url)

Luego podemos usar `.status_code` para saber el código  de estado de la consulta.

In [3]:
print(f'Status: {response.status_code}')

Status: 200


In [4]:
# El output de esta respuesta particular
# puede ser transformado desde JSON a dict
print(response.json()['body'])

 # Issues

Este foro está para que **compartan sus dudas** 🤔. También será usado para **publicar información relevante** en algunos casos. 

Este es un **medio oficial de comunicación** por lo que es tu deber **revisarlo periódicamente**, ojalá todos los días en que trabajes en el curso, para que no estés desinformado/a.


### Reglas del foro 📏:

 1. **Revisar si alguien más ya preguntó por tu duda.** Si se repite algo ya preguntado en otra issue, se cerrará la pregunta repetida y no se responderá. No olvides que también podría estar en alguna issue cerrada. Para revisar las issues puedes apoyarte en el buscador y los compilados de issues 📖.

1. **Googlear antes de crear una issue.** Si tu duda no está en internet, o no te queda claro, crea una issue 🔍.

1. **No puedes postear código** de tus tareas ni una solución directa a algún problema de tus compañeros, relacionado a la tarea 🙅‍♂️. Si tienes una duda puedes crear un código de ejemplo, que no entregue suficiente información, con ta

## Ejemplo consumiendo una API (método `get`)

A continuación vamos a experimentar con la API de [_fakerapi_](https://fakerapi.it/api/v1/). Esta nos genera diferentes tipo clase de información inventada. Por ejemplo, nombre de personas, libros, etc.

En primer lugar, todas las APIs tendrán una **URL BASE**, que consiste en un _link_ de prefijo que toda consulta debe tener. Para el caso de _fakerapi_, esta URL BASE es `"https://fakerapi.it/api/v1/"`. Por lo tanto, cualquier solicitud que hagamos, debe partir con ese prefijo.

Luego, tenemos los **_endpoint_**. Para efectos del curso, podemos considerar estos **_endpoints_** como las diferentes rutas que dispone la API para hacer consultas. Para el caso de _fakerapi_, algunos **_endpoints_** son

* `books/`: nos retornará información de libros inventados.
* `texts/`: nos retornará textos inventados.
* `person/`: nos retornará información personas inventadas.
* `addresses/` nos retornará direcciones inventadas.

Por lo tanto, para consultar esta api, deberemos combinar la **URL BASE** con un **_endpoint_**. De este modo, a URL final sería `"https://fakerapi.it/api/v1/books"`, `"https://fakerapi.it/api/v1/addresses"`, entre otros. Siempre que visitemos una API, esta nos detallará los diferentes _endpoints_ que tiene.

Vamos a probar los primeros 2 **_endpoints_** de _fakerapi_.

In [5]:
import requests

BASE = "https://fakerapi.it/api/v1/"
endpoint_1 = "books/"

solicitud = requests.get(BASE + endpoint_1)

# imprimir el status_code
print(solicitud.status_code)

200


Ahora vamos a ver el contenido de la consulta

In [6]:
data = solicitud.json()
print("Llaves: ", data.keys())

Llaves:  dict_keys(['status', 'code', 'total', 'data'])


In [7]:
data["total"]

10

In [8]:
data['data'][0]

{'id': 1,
 'title': 'HAD THIS FIT--".',
 'author': 'Cindy Weber',
 'genre': 'Ut',
 'description': "SAID was, 'Why is a raven like a telescope! I think I can find it.' And she began fancying the sort of present!' thought Alice. 'I've read that in about half no time! Take your choice!' The Duchess.",
 'isbn': '9795515783807',
 'image': 'http://placeimg.com/480/640/any',
 'published': '2017-09-16',
 'publisher': 'Aut Repudiandae'}

Cómo podemos notar, nos inventó el nombre, género y descripción de un libro. Ahora vamos a hacer una consulta al endpoint de `"texts/"`.

In [9]:
endpoint_2 = "texts/"

solicitud = requests.get(BASE + endpoint_2)

data = solicitud.json()
data['data'][:2]

[{'title': "Alice. 'I've read.",
  'author': 'Emilio Bosco',
  'genre': 'Voluptatem',
  'content': "Bill, the Lizard) could not help thinking there MUST be more to do that,' said the King, 'that saves a world of trouble, you know, upon the other paw, 'lives a March Hare. 'Exactly so,' said Alice."},
 {'title': 'The Antipathies, I.',
  'author': 'Zakary Rogahn',
  'genre': 'Soluta',
  'content': "Tortoise--' 'Why did they draw the treacle from?' 'You can draw water out of a good deal: this fireplace is narrow, to be no doubt that it led into the garden, where Alice could only see her. She is."}]

### Parámetros de una consulta

Cada **_endpoint_** puede tener sus parámetros para personalizar la consulta. 

Por un lado, _fakerapi_ tiene algunos parámetros que sirve para cualquier **_endpoint_**. Uno de estos es `"_quantity"`, y nos permite indicar la cantidad de elementos a inventar. Por defecto es 10, pero ahora vamos a cambiarlo a 3. 

Por otro lado, el **_endpoint_** de `texts/` tiene un parámetro específico llamado `"_characters"`. Este permite indicar la cantidad de caracteres a utilizar en el texto que inventará. Por defecto es 200, pero ahora vamos a cambiarlo a 20.

In [10]:
endpoint_2 = "texts/"

parametros = {"_quantity": 3, "_characters": 20}
solicitud = requests.get(BASE + endpoint_2, params=parametros)

data = solicitud.json()
data['data']

[{'title': 'Alice, and tried.',
  'author': 'Felicita Heaney',
  'genre': 'Possimus',
  'content': 'It did so indeed.'},
 {'title': 'I must sugar my.',
  'author': 'Berta Hamill',
  'genre': 'Corrupti',
  'content': "Cheshire Cat,'."},
 {'title': 'MORE than.',
  'author': 'Frederique Waelchi',
  'genre': 'Doloremque',
  'content': 'March Hare and the.'}]

### Retorno de una API

Así como esta API, existen muchas otras. No todas responden en formato JSON.


In [11]:
# Podemos usar una API para obtener nuestra IP pública
# Notar que no estamos transformando a JSON
response = requests.get('https://api.ipify.org')
ip = response.text
print(response.status_code, ip)

200 190.161.94.143


In [12]:
# Podemos ahora usar una API para obtener la latitud y 
# longitud en la que nos encontramos al momento de correr este código
response = requests.get(f'https://ipapi.co/{ip}/latlong/')
print(response.status_code, response.text)

200 -33.452100,-70.653600


Por temas de seguridad, muchas de las APIs públicas necesitan una llave/clave para poder utilizarlas. Para conseguir estas *keys* en general debes crearte una cuenta. De esta forma se mantiene control de la aplicación expuesta, quiénes están accediendo a ella, con qué frecuencia, etc.


In [13]:
# Además, podemos usar otra API para ver más descripciones de la IP utilizada
url = 'http://api.ipstack.com/'
# En este caso puedes usar esta `API_KEY` para probar
API_KEY = 'c657ed216cf3e05d129bd6b2ccb8589e'
# Recibe la API_KEY como parámetro

# Esto puede ser enviado de dos formas:

# 1. Agregando los parámetros en la URL:
pais = requests.get('{}/{}?access_key={}'.format(url, ip, API_KEY))

# 2. Pasando los parámetros en el método:
pais = requests.get(f'{url}/{ip}', params={'access_key': API_KEY})

In [14]:
pais.status_code

200

In [15]:
pais.headers

{'content-type': 'application/json', 'transfer-encoding': 'chunked', 'date': 'Fri, 09 Jun 2023 20:13:37 GMT', 'x-apilayer-transaction-id': '5088d27d-48ac-4a3c-865e-e5d9fd399aa3', 'access-control-allow-origin': '*', 'access-control-allow-methods': 'GET, POST, HEAD, OPTIONS', 'access-control-allow-headers': '*', 'x-quota-limit': '100', 'x-quota-remaining': '93', 'x-increment-usage': '1', 'x-request-time': '0.029'}

In [16]:
# En este caso la API sí retorna un JSON
pais.json()

{'ip': '190.161.94.143',
 'type': 'ipv4',
 'continent_code': 'SA',
 'continent_name': 'South America',
 'country_code': 'CL',
 'country_name': 'Chile',
 'region_code': 'RM',
 'region_name': 'Santiago Metropolitan',
 'city': 'Santiago',
 'zip': None,
 'latitude': -33.46500015258789,
 'longitude': -70.65599822998047,
 'location': {'geoname_id': 3871336,
  'capital': 'Santiago',
  'languages': [{'code': 'es', 'name': 'Spanish', 'native': 'Español'}],
  'country_flag': 'https://assets.ipstack.com/flags/cl.svg',
  'country_flag_emoji': '🇨🇱',
  'country_flag_emoji_unicode': 'U+1F1E8 U+1F1F1',
  'calling_code': '56',
  'is_eu': False}}

### Uso de `post`

En el caso de la API con la que hemos estado haciendo estas pruebas, solo se ofrecen servicios para realizar consultas, lo que se puede llevar a cabo utilizando el método `GET` del protocolo HTTP. Sin embargo, muchas veces queremos crear recursos en nuestro servidor, como por ejemplo crear un nuevo artículo para un *blog*, y para esto debemos utilizar el método `POST` del protocolo.

La API de `JSONPlaceholder` nos permite simular el uso de una API real, sin que verdaderamente exista un servicio detrás de esta. En este caso la utilizaremos para simular la creación de un artículo para un *blog*. En la práctica no estará ocurriendo nada en el servidor, puesto que es solo una simulación, pero en la vida real uno esperaría que como respuesta a nuestra *request* se cree una entrada en la base de datos del servicio que estamos utilizando. Pueden ver más información de como usar esta API [aquí](https://jsonplaceholder.typicode.com/).

A diferencia del método `GET`, cuando utilizamos el método `POST` podemos enviar información a la API utilizando el parámetro `data`, al cual podemos pasarle un diccionario de Python con la información que queremos enviar. Para este ejemplo debemos enviar la información de un artículo noticioso que queremos crear.


In [17]:
# Extracto de un artículo proveniente del blog
# https://venezolanoenchile.com/2016/01/20/como-es-el-clima-en-santiago-de-chile/
cuerpo = '''
El clima de Santiago es muy extraño para los que venimos de un país tropical
como Venezuela y, más aún, para los que vivíamos en Maracaibo, como yo.

Un día de verano, mientras caminaba a eso de las 2pm hacia mi trabajo, me puse
a pensar en los temas que no he tocado aún en el blog. En ese momento, con 32°C
de temperatura y bajo el sol, decidí escribir sobre este tema.

Chile es un país que tiene muchos tipos de clima, desde el desértico hasta el
frío antártico. Pero como yo no conozco ninguna otra ciudad de Chile que no sea
Santiago, todo lo que diré a continuación será de la capital.'
'''

data = {
    'title': '¿Cómo es el clima en Santiago de Chile?',
    'body': cuerpo,
    'userId': 1,
}

noticia = requests.post('https://jsonplaceholder.typicode.com/posts', data=data)

In [18]:
# Vemos que obtenemos un código de que nuestro artículo fue creado
print(noticia.status_code)
print(noticia.reason)

# Esta API nos retorna un JSON con el mismo recurso creado,
# nótese que se le asignó un id al artículo
data = noticia.json() 
print(data.keys())

201
Created
dict_keys(['title', 'body', 'userId', 'id'])


In [19]:
print(str(data['id']) + "\n")
print(data['title'] + "\n")
print(data['body'])

101

¿Cómo es el clima en Santiago de Chile?


El clima de Santiago es muy extraño para los que venimos de un país tropical
como Venezuela y, más aún, para los que vivíamos en Maracaibo, como yo.

Un día de verano, mientras caminaba a eso de las 2pm hacia mi trabajo, me puse
a pensar en los temas que no he tocado aún en el blog. En ese momento, con 32°C
de temperatura y bajo el sol, decidí escribir sobre este tema.

Chile es un país que tiene muchos tipos de clima, desde el desértico hasta el
frío antártico. Pero como yo no conozco ninguna otra ciudad de Chile que no sea
Santiago, todo lo que diré a continuación será de la capital.'



### Autenticación en _headers_

Aparte del método `post`, también tenemos `.put()`, `.patch()` o `.delete()`. Generalmente, este tipo de requests repercuten en la modificación de una base de datos, y para lograr esto, muchas veces es necesario tener una autorización previa. 

Es aquí que surge la necesidad de ocupar el parámetro `headers` que permite incluir información en la cabecera de la solicitud incluyendo, por ejemplo, un _token_ especial de acceso que a la API le sirve para identificarte y verificar si tienes los permisos. 

Utilizar este parámetro es análogo a `data`. Solo es necesario crear un diccionario e incluirlo en la _requests_. Las llaves del diccionario puede cambiar según la API que queramos acceder.

In [20]:
my_headers = {
    "Authorization": "MI-TOKEN"
}

data = {
    'title': 'Probando',
    'body': 'Probando',
    'userId': 11,
}

requests.post('https://jsonplaceholder.typicode.com/posts', headers=my_headers, data=data)

<Response [201]>

### Caso aplicado: API de GITHUB

A continuación, te presentamos un caso aplicado para crear una _issue_ en un repositorio de Github mediante su API.  En este caso, la API de Github pide que el `data` a enviar sea un diccionario en su forma de _string_, es decir, `json.dumps(data)`.

Además, para que este caso aplicado funcione correctamante, se requiere:

1. Completar `github_repo` con el nombre de tu repositorio de github dado en el curso. Por ejemplo: `"juanito-iic2233-20XX-1"`
2. Completar `token` con un _access token_ generado en Github. Puedes generar uno en [esta página](https://github.com/settings/tokens/new). Este _token_ debe tener al menos el siguiente permiso para poder crear una _issue_: _Full control of private repositories_.

In [ ]:
import json

github_repo = 'juanito-iic2233-20XX-1'
token = "COMPLETAR"

body = {
    'title': "Creando una issue con la API",
    'body': "Ahora tengo el poder para hacer issues desde Python! 🎉"
}

my_headers = {
    'Authorization': 'token ' + token,
    'Accept': 'application/vnd.github.v3+json'
}

url = f"https://api.github.com/repos/IIC2233/{github_repo}/issues"

response = requests.post(url, data=json.dumps(body), headers=my_headers)
response.status_code

Si la celda anterior arrojó un `201`. Significa que el _token_ utilizado y el `github_repo` estaban correctos, así que se creó la _issue_. Ejecuta la siguiente celda para ver el _link_ a dicha issue.

In [ ]:
response.json()["html_url"]

## *Server-side App*

La misión principal del servidor es disponer el contenido para que pueda ser consultado mediante un *web service*. La aplicación que corre en el servidor es la encargada de la lógica e interacción entre cliente-servidor. La información que viaja entre un cliente y un servidor permite generar comunicación entre aplicaciones.

Una aplicación puede estar desarrollada en cualquier lenguaje de programación que permita exponer una API para ser consumida por otras aplicaciones a través de la web. Por ejemplo, podemos tener una aplicación corriendo en Java, y desde nuestro código en Python acceder a esa API.

En Python existe [WSGI](https://docs.python.org/es/3/library/wsgiref.html) para exponer APIs. Tambien existen varios *frameworks* de programación que facilitan esta misma tarea, como **Flask** y **Django**. Además, puedes montar tus aplicaciones en servicios o servidores ya disponibles en la web, provistos como Platform-as-a-Service (PaaS) o Infrastructure-as-a-Service (IaaS). Por ejemplo, puedes usar **Heroku** (PaaS), **Digital Ocean** (IaaS), o **Microsoft Azure** (PaaS) para disponer tus APIs en una red pública con alta disponibilidad.


A modo de ejemplo, tenemos una pequeña API utilizando WSGI para levantar una aplicación que responde algunos mensajes.

Antes de empezar, debes ejecutar el archivo `servidor.py` en la terminal o en VSCode para levantar esta API. Luego, puedes ejecutar las siguientes celdas donde utilizaremos `requests` para realizar solicitudes a esta API.

In [ ]:
# La URL base de la API creada
BASE_URL = "http://localhost:4444/"

# Podemos consultar a esta ruta
solicitud = requests.get(BASE_URL)
solicitud.status_code

200

In [ ]:
solicitud.json()

{'message': 'Hello World'}

In [ ]:
# ¡Vamos a despedirnos haciendo una consulta a otro endpoint de nuestra API!
respuesta = requests.get(BASE_URL + "goodbye/")
print(respuesta.status_code, respuesta.json())

200 {'message': 'Que la fuerza esté contigo'}


**Te invitamos a consumir diferentes APIs desde Python para poner a prueba este contenidos**. Algunas ideas de estudio pueden ser:
- Revisar los diferentes _endpoint_ que tenía [_fakerapi_](https://fakerapi.it/en) para inventar más información.
- Revisar la documentación de [OMDb API](http://www.omdbapi.com/#usage) para buscar películas.
- Usar la APi de [pokémon](https://pokeapi.co/docs/v2#pokemon) para buscar información de tu pokémon favorito.
